haystack/nodes/prompt/prompt_node.py

```diff
@@ -358,6 +360,9 @@ def run(
         if "prompt_template" not in invocation_context.keys():
             invocation_context["prompt_template"] = self.get_prompt_template(prompt_template)
 
+        if generation_kwargs and "generation_kwargs" not in invocation_context.keys():
+            invocation_context["generation_kwargs"] = generation_kwargs
```

Comment:
I think it would be better if we actually merged this two. The `invocation_context` is passed as kwargs to `PromptNode.prompt_model.invoke()` in the end.

Making this change wouldn't require us to modify any existing and future invocation layer to handle custom kwargs, it's much more maintanable.

Reply:
from the result and code simplicity point of view, the 2 params can indeed be merged; however, the meaning of these two parameters are different, `invocation_context` refers to all the parameters passed to the invoke function of the following layers, and `generation_kwargs` refers to the parameters that affect the performance of the model generation alone. Therefore, `generation_kwargs` should be a subset of `invocation_context`. Furthermore, `invocation_context` is the output of the Shaper, if there is a Shaper node in the pipeline. In this case, the invocation_context can't be passed as an argument, e.g. for this code snippet:

```python
result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
    "Prompter": { "prompt_template": "question-answering", "invocation_context": {"generation_kwargs": {"do_sample": True}}},
    "Retriever": {"top_k": 1},
    "Reranker": {"top_k": 1}
    })
```

Below is the error log for 
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/haystack/pipelines/base.py", line 552, in run
    node_output, stream_id = self._run_node(node_id, node_input)
  File "/usr/local/lib/python3.8/dist-packages/haystack/pipelines/base.py", line 467, in _run_node
    return self.graph.nodes[node_id]["component"]._dispatch_run(**node_input)
  File "/usr/local/lib/python3.8/dist-packages/haystack/nodes/base.py", line 201, in _dispatch_run
    return self._dispatch_run_general(self.run, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/haystack/nodes/base.py", line 245, in _dispatch_run_general
    output, stream = run_method(**run_inputs, **run_params)
TypeError: run() got multiple values for keyword argument 'invocation_context'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "run_pipeline.py", line 18, in <module>
    result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
  File "/usr/local/lib/python3.8/dist-packages/haystack/pipelines/base.py", line 559, in run
    raise Exception(
Exception: Exception while running node 'Prompter': run() got multiple values for keyword argument 'invocation_context'
Enable debug logging to see the data that was passed when the pipeline failed.
```

Reply:
Oh no, I didn't mean to change the naming of the `PromptNode` arg. What I mean is change the name of the param to `invocation_kwargs` as I said in the other comment and here instead of adding it to `invocation_context` as a new field we merge them together. 

Something like this.

```suggestion
        invocation_context.update(invocation_kwargs)
```

This way we don't need to change anything else, nor in the `PromptNode` nor in any invocation layer.

Reply:
I didn't give it a try to be fair, but I'd expect this to work.

Reply:
I see your point. But this approach would be different from what @vblagoje suggested earlier. Can we have a quick chat in Discord? I have sent you guys friend requests. That will be more efficient.

Reply:
Ok, I'm ok with the naming. 

But what's stopping us from just merging `generation_kwargs` with `invocation_context` if in the end we just merge it with `invoke`'s `kwargs`?

The `invocation_context` is exploded as `PromptNode.__call__()`'s `kwargs` in line [366](https://github.com/deepset-ai/haystack/pull/4832/files#diff-977894d453c60a383b8c831ae90d32076823f4082a3ea5e9792054b33299929eR366), then again as `kwargs` to `PromptNode.prompt()` and then end up in [`PromptModel.invoke` as `kwargs_copy`](https://github.com/deepset-ai/haystack/blob/1dfb7454bfbe20a7c6f6947163cd7955725ea89b/haystack/nodes/prompt/prompt_node.py#LL174C68-L174C68) that just proxies them to the `InvocationLayer` used.

Reply:
Yeah by all means - if possible why not!

Reply:
> Ok, I'm ok with the naming.
> 
> But what's stopping us from just merging `generation_kwargs` with `invocation_context` if in the end we just merge it with `invoke`'s `kwargs`?
> 
> The `invocation_context` is exploded as `PromptNode.__call__()`'s `kwargs` in line [366](https://github.com/deepset-ai/haystack/pull/4832/files#diff-977894d453c60a383b8c831ae90d32076823f4082a3ea5e9792054b33299929eR366), then again as `kwargs` to `PromptNode.prompt()` and then end up in [`PromptModel.invoke` as `kwargs_copy`](https://github.com/deepset-ai/haystack/blob/1dfb7454bfbe20a7c6f6947163cd7955725ea89b/haystack/nodes/prompt/prompt_node.py#LL174C68-L174C68) that just proxies them to the `InvocationLayer` used.

the only problem I see with this approach is that we need to list all possible params of generation_kwargs in the line 183 of hugging_face.py. But compared to the fact that we have to update all providers' invocation layers, this approach has much less code change and is easier to maintain. So I updated the PR, pls review it. Thanks a lot!